### ISSUE TITLE
setup.py

### ISSUE THREAD

**Issue by noraj** [jtesta/ssh-audit#46]
```
$ cd pypi
$ python setup.py build
Traceback (most recent call last):
  File "setup.py", line 8, in <module>
    version = re.search('^VERSION\s*=\s*\'v(\d\.\d\.\d)\'', open('sshaudit/sshaudit.py').read(), re.M).group(1)
FileNotFoundError: [Errno 2] No such file or directory: 'sshaudit/sshaudit.py'

$ python setup.py build
Traceback (most recent call last):
  File "setup.py", line 11, in <module>
    with open("sshaudit/README.md", "rb") as f:
FileNotFoundError: [Errno 2] No such file or directory: 'sshaudit/README.md'
```

There is no `sshaudit.py` in `pypi/sshaudit/`, I suspect you are using a symbolic link locally.
Also it's not the right filename (dash missing). Same for the README.

So the workaround was:

```
$ cp ../ssh-audit.py sshaudit/sshaudit.py
$ cp ../README.md sshaudit/README.md
```

**Comment by jtesta** [comment]
In the dev branch, see the packages/notes.txt file for instructions on 
how to build a package.  I just checked in some minor changes so it'll 
work on a development build.

Out of curiosity, why are you trying to build a PyPI package yourself?

**Comment by jugmac00** [comment]
Hi @jtesta 

this is a topic, I planned to talk about for quite a time, but I was quite busy recently.

Actually, the `setup.py` is not only for creating a package for PyPi, but also for installing it locally and some more things.

To be honest, this package here is the very first Python repository I ever encountered, where the `setup.py` is not in the project's root folder.

This is nothing especially problematic, but it is super unusual.

There are many tools out there which rely on `setup.py` being in the root of the project, which help preventing errors or help with testing, linting and more.

e.g. `check-python-version` makes sure that you put the same Python versions in your `setup.py`, in your Travis config, in your tox config and so on.... And there are many more helper tools out there.

As @noraj indicated, in order to leverage my usual tooling, I created a symlink into the project's root.

Actually, while fixing the test / tox setup and fixing / creating the type annotations, I noticed quite a few `oddities` which I do not see in other Python projects.

I hesitated to report them all at once, as first and foremost this project does an excellent job, and aligning it to Python standards is not a top priority, I guess. Also I did not want to report them, when I had no time to work on them.

I will try to create the issues either tonight, or beginning of next week. And rest assured - its not that big of an issue, and certainly, I will assist you as much as I can.

**Comment by jtesta** [comment]
When I created this fork, there was no setup.py.  I created it myself 
when I needed to make a PyPI package.  I remember running into errors 
that didn't make sense, so I just tried random things until I stumbled 
on something that worked.  At that point, I crossed this off my list and 
moved on to other things.  So that explains why things are so... 
non-standard.

I'm certainly open to fixing things up!

**Comment by noraj** [comment]
> Out of curiosity, why are you trying to build a PyPI package yourself?

1. The generic answer is: security. It's always better to build from source oneself than trusting pre-built binaries.
2. On my specific case it's because I'm [BlackArch](https://blackarch.org/) dev and I upgraded the ssh-audit package from the original repo to your fork. See https://github.com/BlackArch/blackarch/pull/2803/files

**Comment by jtesta** [comment]
>  1. The generic answer is: security. It's always better to build from
>     source oneself than trusting pre-built binaries.

I'd strongly disagree.

Many, many moons ago, I ran the Beyond Linux From Scratch "distro" on my 
main machine (if you're not familiar, it's nothing more than a large set 
of instructions to compile *everything* yourself from scratch!)  After a 
couple weeks, my machine was riddled with security holes.  It quickly 
became impossible for me to keep up on re-compiling all the embedded 
libraries that had security updates.  For example, when OpenSSL or zlib 
had an update, I had to spend 20-60 minutes compiling, debugging errors, 
then researching what software statically linked to them, THEN 
re-compile (and maybe debug) those as well.  I couldn't keep up.

BLFS was great at teaching me how to debug configuration & compilation 
errors.  It was an excellent learning experience.  But it also 
inadvertently taught me that using supported packages with automatic 
patching was an absolute necessity for any usable system.  Now I'll bend 
over backwards to use a supported package instead of sources!


>  2. On my specific case it's because I'm BlackArch
>     <https://blackarch.org/> dev and I upgraded the ssh-audit package
>     from the original repo to your fork. See
>     https://github.com/BlackArch/blackarch/pull/2803/files

Ahh, gotcha.  Well in that case, I wouldn't package the dev branch. 
It's got some experimental features in it, which should get polished up 
in the next week or two for a release.  I'm looking forward to it!

**Comment by jugmac00** [comment]
I suggest the following changes:
- move `setup.py` to root directory
- move `MANIFEST.in` to root directory
- ... ie. move all files of `packages` folder to the root
(if you think there are too many files in the root then... usually you put your source code into a `src` folder, like here https://github.com/jugmac00/flask-reuploaded - I would not do this all at once.. maybe at a later time, as this would mean a bit more work and some explanation how this works)

Having the configuration files in the project root is Python standard, and allows to leverage the following tools (for each which I'd setup a separate tox env):
- https://pypi.org/project/check-python-versions/
- https://pypi.org/project/check-manifest/

May or may not related to your experienced problems with creating a package, is the actual file name of the main module `ssh-audit.py`.

While not technically illegal, the 99,99999% :-) case is that Python modules are always lower cased, one word, or when more words **separated with an underscore**.

**Update**
https://github.com/jtesta/ssh-audit/blob/8e3f3c60448279ab03a378d3961ad212afb08eab/packages/Makefile.pypi#L2
Gotcha :-)

So I'd suggest to rename `ssh-audit.py` (and only this file) to `ssh_audit.py`, so standard tools and also standard import is working.

Ever tried to fire up a Python interpreter and `from ssh-audit import AuditConf`? It will break with a syntax error. Use case for this is (easier) testing, and reusing code.

This non-standard naming also required the previous maintainer to use special syntax to get hold of the program for testing:
https://github.com/jtesta/ssh-audit/blob/8e3f3c60448279ab03a378d3961ad212afb08eab/test/conftest.py#L8-L12

To be clear - I only suggest to rename the module (file), not the repo, not the project name, and also not the entry point you defined in `setup.py` - this means, after installation, the program is still available as `ssh-audit`.
https://github.com/jtesta/ssh-audit/blob/8e3f3c60448279ab03a378d3961ad212afb08eab/packages/setup.py#L31

This also would be a mandatory change if you ever plan to split your module into smaller pieces. These 3.5 k lines are starting to get a bit unwieldy :-) If you consider this last point, it probably should be done after the next release - there'd be quite some restructuring steps involved. (You can have a look at my password checker cli https://github.com/jugmac00/hibpcli/tree/master/src/hibpcli - way not that elaborate as `ssh-audit` but you get the idea).

If you wonder how to do a PyPi release then.. I just created a step by step guide for my latest project: 
https://github.com/jugmac00/flask-reuploaded/blob/master/PACKAGING.rst

This is based on the there linked article
https://hynek.me/articles/sharing-your-labor-of-love-pypi-quick-and-dirty/


**TL/DR**
- move all files, currently under packaging into the project root (except this `sshaudit` subfolder, which is then no longer necessary)
- rename `ssh-audit.py` to `ssh_audit.py`
- rename `notes.txt` into e.g. `PACKAGING.md`
- update makefiles

**Possible future steps**
- move source code into `src` directory
- split main module into smaller ones


If you are ok, I'd create a PR in the next couple of days.

**Comment by jtesta** [comment]
> - move all files, currently under packaging into the project root (except this sshaudit subfolder, which is then no longer necessary)
> - rename ssh-audit.py to ssh_audit.py
> - rename notes.txt into e.g. PACKAGING.md
> - update makefiles

Yep, all that sounds good.  After the upcoming v2.3.0 release is done, we can get started on that (I'm hoping it'll be done by the end of this week).  The only other thing I'd like to note is that ssh-audit.py should remain as a stub which imports all the code from ssh_audit.py.  That way users can still run the program like they did before without installing it ("./ssh-audit.py target").

> - move source code into src directory

Since this isn't standard Python practice, we don't need to do it.  I'd rather re-organize the project to adhere to Python conventions.

> - split main module into smaller ones

Yes, I'd _love_ to do this.  I've been wanting to do it for a while, but with all the other problems we had (broken tests, tabs instead of spaces, etc), I pushed it off.  Now we're in the position to finally do it!

I'll get this going myself after we fix the other issues, above.

**Comment by jugmac00** [comment]
> Yep, all that sounds good. After the upcoming v2.3.0 release is done, we can get started on that (I'm hoping it'll be done by the end of this week). The only other thing I'd like to note is that ssh-audit.py should remain as a stub which imports all the code from ssh_audit.py. That way users can still run the program like they did before without installing it ("./ssh-audit.py target").

I bet you know, but `ssh-audit` will be still available after installation.

Also, there is a nifty trick I saw at another cli app. You can create a bash wrapper for the Python script, so you can run `./app` without installing it and still use a `src` directory which... see next paragraph :-)
https://github.com/mgedmin/check-python-versions/blob/master/check-python-versions

(But let's do not hide this in the `ssh-audit.py` file :-))


> > * move source code into src directory
> Since this isn't standard Python practice, we don't need to do it. I'd rather re-organize the project to adhere to Python conventions.

Actually, it is an emerging standard, which is followed by tons and tons of projects, and many older projects convert to it.

Python is a very evolving language, and especially the packaging story was a tough one, but it is getting better and better.

Just a few projects which converted to the `src` strategy...
https://github.com/python-attrs/attrs
https://github.com/pallets/flask
almost all Zope projects (hundreds)
https://github.com/getsentry/sentry
https://github.com/python-pillow/Pillow
https://github.com/pypa/pip
https://github.com/pytest-dev/pytest
https://github.com/twisted/twisted
https://github.com/Pylons/pyramid
https://github.com/pyca/cryptography
...

There many good reasons... some are outlined in the following articles.... btw both Hynek and Paul are `cpython` core developers, and Paul is even a member of the Python packaging authority. So it is not **somebody** on the internet :-)

https://hynek.me/articles/testing-packaging/
https://web.archive.org/web/20170615032800/https://enotuniq.org/
https://blog.ganssle.io/articles/2019/08/test-as-installed.html


> > * split main module into smaller ones
> 
> Yes, I'd _love_ to do this. I've been wanting to do it for a while, but with all the other problems we had (broken tests, tabs instead of spaces, etc), I pushed it off. Now we're in the position to finally do it!

For this to become a smooth experience, a `src` directory is (almost) mandatory.

> I'll get this going myself after we fix the other issues, above.

There is no need to rush anything. I still think a one time more or less breaking change (`ssh-audit.py` -> bash wrapper `ssh-audit`) could enable a very convenient future development of the application.

My background is not security - I maintain several mid sized, and one huge Python application (all proprietary) - that is certainly a different story compared to this application here. Though, a straight project layout will make it easier in the future.

If you ever want to chat about any Python related topics, I am sure we'll find a time slot which both fits us (US vs EU based).
